Make quest invite modal pop up on user sync, challenge modal display before invite modal, and additional party tweaks

hairlessbear opened the pull request:
- Added a Boolean `invited` field to the user model that defaults to `false`
- Set `invited` to true for all group members (other than quest leader) when quest invitation is sent ([code link](https://github.com/HabitRPG/habitrpg/blob/0fa93506f5ea2eb68284486dc0a37bd970c1e7a3/website/src/controllers/groups.js#L915))
- Set `invited` to false when:
  - Quest is cancelled before it starts (all group members; [code link](https://github.com/HabitRPG/habitrpg/blob/0fa93506f5ea2eb68284486dc0a37bd970c1e7a3/website/src/controllers/groups.js#L980-982))
  - User accepts the invitation (single user; [code link](https://github.com/HabitRPG/habitrpg/blob/0fa93506f5ea2eb68284486dc0a37bd970c1e7a3/website/src/controllers/groups.js#L951))
  - User rejects the invitation (single user; [code link](https://github.com/HabitRPG/habitrpg/blob/0fa93506f5ea2eb68284486dc0a37bd970c1e7a3/website/src/controllers/groups.js#L962))
  - Quest starts (all group members; [code link](https://github.com/HabitRPG/habitrpg/blob/0fa93506f5ea2eb68284486dc0a37bd970c1e7a3/website/src/models/group.js#L145), function called [here](https://github.com/HabitRPG/habitrpg/blob/0fa93506f5ea2eb68284486dc0a37bd970c1e7a3/website/src/controllers/groups.js#L835))
- Replaced the quest invite modal notification logic with `user.party.quest.invited == true`, causing the invite modal to popup on user sync without causing overhead issues as discussed in [#5239](https://github.com/HabitRPG/habitrpg/pull/5239#issuecomment-103699467).

Everything is working as intended, except for the rewards display. The quest invite modal is currently displaying like this:

![selection_022](https://cloud.githubusercontent.com/assets/7059890/7828154/b7452faa-03fd-11e5-8fde-68c4a657bc1e.png)

...and I can't figure out why. I'm guessing it has something to do with the scope of the [`quest-rewards`](https://github.com/HabitRPG/habitrpg/blob/develop/common/script/public/directives.js#L155-164) directive, but I'm really not sure. pinging @crookedneighbor and @Alys for help.
### EDIT by Blade: Still needs to be done
- [ ] Migration script to add `party.quest.RSVPNeeded = true` to users who have pending quests
- [x] Tests around watch function
### Would be nice to do
- [x] Accept and reject buttons on party page (migration script less necessary)
- [x] Party sync when party page loads
- [x] ~~Refactor fetch recent messages~~ EDIT by hairlessbear: to be done in a later PR
- [x] Quest completion modal should appear before quest invite modal

crookedneighbor commented on website/src/models/user.js:
```diff
@@ -280,7 +280,8 @@ var UserSchema = new Schema({
       },
-      completed: String // When quest is done, we move it from key => completed, and it's a one-time flag (for modal) that they unset by clicking "ok" in browser
+      completed: String, // When quest is done, we move it from key => completed, and it's a one-time flag (for modal) that they unset by clicking "ok" in browser
+      invited: {type: Boolean, 'default': false} // Set to true when invite is pending, set to false when quest invite is accepted or rejected
```
I feel like the name of this should be something a little more semantically meaningful, because even if the user declines, they were still invited, so it doesn't make sense to set `invited` to false. Maybe something like `RSVPNeeded`?

hairlessbear replied:
That's a much better name. `invited` made sense to me, but `RSVPNeeded` is way better. I'll change that right now.

crookedneighbor commented on website/src/models/group.js:
```diff
@@ -141,7 +141,8 @@ var cleanQuestProgress = function(merge){
     },
-    completed: null
+    completed: null,
+    invited: false
```
Is this used any where?

hairlessbear replied:
Yes. It's used [here](https://github.com/HabitRPG/habitrpg/blob/0fa93506f5ea2eb68284486dc0a37bd970c1e7a3/website/src/controllers/groups.js#L835) when the quest is started, and takes care of any member who didn't accept or reject the quest.

crookedneighbor commented:
I feel like I had the same problem, but I don't remember what I did to fix it. I'll play with it and get back to you.

In the meantime, something you've done has caused the push notification test for receiving a quest invite to fail. That'll need to be sorted out before this can be merged.

crookedneighbor commented:
Oh I see what the test problem is regarding the test. I was mocking the User model in the test, and you added a call to `User.update`, which was not being mocked .

crookedneighbor commented on website/public/js/controllers/notificationCtrl.js:
```diff
@@ -148,8 +148,9 @@ habitrpg.controller('NotificationCtrl',
-      if (after == before || after != true) return;
+    $rootScope.$watch('user.party.quest.invited == true', function(after, before){
+      if (after != true) return;
+      $rootScope.party.$get();
```
Angular is throwing an error on this.

crookedneighbor commented on website/public/js/controllers/notificationCtrl.js:
```diff
@@ -148,8 +148,9 @@ habitrpg.controller('NotificationCtrl',
     // Quest invitation modal
-    $rootScope.$watch('party.quest.key && !party.quest.active && party.quest.members[user._id] == undefined', function(after, before){
-      if (after == before || after != true) return;
+    $rootScope.$watch('user.party.quest.RSVPNeeded == true', function(after, before){
```
Figured out the problem. Change the condition to: `party.quest.key && user.party.quest.RSVPNeeded == true`, that'll put the party.quest.key in scope so the directive can load it.

crookedneighbor replied:
Also, is it necessary to explicitly check for `RSVPNeeded == true`? The two states should be `true` or `false`, so just checking `party.quest.key && user.party.quest.RSVPNeeded` should be fine.

hairlessbear replied:
Adding `party.quest.key` brings us back to the problem this is trying to fix, and the modal won't show up until the page is reloaded (or `fetch recent messages` is pressed on the party page).

crookedneighbor replied:
Good point. Will have to experiment with it.

crookedneighbor commented on website/public/js/controllers/notificationCtrl.js:
```diff
@@ -148,8 +148,9 @@ habitrpg.controller('NotificationCtrl',
-      if (after == before || after != true) return;
+    $rootScope.$watch('user.party.quest.RSVPNeeded == true', function(after, before){
+      if (after != true) return;
+      $rootScope.party.$get();
```
This no longer throws an error when `party.quest.key` is added to the watch check.

crookedneighbor commented:
All that's left is to fix the api test:

Change userMock on line 92 of test/api/pushNotifications.coffee to:

``` coffeescript
userMock = {
  findById: (arg, cb) ->
    cb(null, recipient)
  find: (arg, arg2, cb) ->
    cb(null, [recipient])
  update: (arg, arg2) ->
    { exec: -> true }
}
```

And it'd probably be good to write some tests around the watch, but that can be a tad confusing, so I'll write those wants this is merge ready.

hairlessbear commented:
I figured out a hacky fix to the `quest-rewards` problem. I'm setting `user.party.quest.key` when an invitation is sent, and using that instead of `party.quest.key` when calling `quest-rewards` on the invite modal. 

Only one issue left now: when a user accepts or rejects the quest, if you refresh the page, the invite modal pops up again. But, if you refresh the page a second time, it doesn't show up. I'm guessing there's some sort of issue with `User.update`, but I'm not sure what yet. After `accept` or `reject` are clicked, the values in mongoDB are being updated properly, so I guess they aren't being updated in the browser or something...? I'm investigating now.

crookedneighbor commented:
Haven't, looked at it yet, but you may need to call user sync when the user makes a decision to force the browser to be in sync with the server. 

Probably what is happening is you refresh, the browser's version of the data says the RSVPNeeded is true and then the modal pops up. The next action is for the browser to sync with the server, which sets RSVPNeeded correctly. Just a theory. Does that make sense?

The trick will be to sync AFTER the server is done saving.

hairlessbear commented:
The problem only occurs if the user refreshes immediately after clicking accept/reject without doing anything else. As soon as anything is done to cause user sync (e.g., check off a task), refreshing doesn't do anything. The easiest solution would be to force a user sync after each update call; I'll try to get that working.

hairlessbear commented:
Did we just comment simultaneously saying almost the same thing? That's amusing :laughing:

crookedneighbor commented:
I have an idea. Are all your most recent commits pushed?

hairlessbear commented:
Almost. I updated all of the items in the invite modal to use `user.party.quest.key` for consistency, and changed the notification to not explicitly check for `true`. I'll push both right now.

hairlessbear commented:
Oh, and I removed `$rootScope.party.$get` from the watch because it's no longer necessary (it was there for `party.quest.key` in the invite modal). Feel free to put it back if I was wrong in removing it.

crookedneighbor commented:
@hairlessbear Well, the good news is I did get it to sync after calling party.$questAccept and $party.questReject by utilizing that those functions are promises, so I can call User.sync() after the request to the server resolves: https://github.com/HabitRPG/habitrpg/commit/42b482cb7efa8b37941e9fa0a58bba7a7bf3a4c6

Bad news: I'm not super great at promises and I don't know how to effectively mock and test them. Will do some research about how best to do that. 

I'd also like to explore attaching that User.sync() function on the end of all these promises: https://github.com/HabitRPG/habitrpg/blob/develop/website/public/js/services/groupServices.js#L34-L37 which would eliminate the need to create wrapper functions. Not sure if that's easy to do ,or even possible, in the current framework.

hairlessbear commented:
I've never heard of promises, so I'm off to Google to learn. I'm glad you got it working though!

We definitely need to figure out a way to sync on `questCancel`. I confirmed via testing that the same refresh issue occurs if a quest is cancelled. It only happens when multiple actions are performed by multiple users in a specific order, so it would likely almost never occur in production, but obviously we still need to fix it.

hairlessbear commented:
Closed due to being superseded by #5305

crookedneighbor commented:
FYI, this or would have been merged in automatically when the other one gets merged as it contains all the commits in this one. GitHub Magic!

hairlessbear commented:
Does that mean I need to reopen it, or will the magic happen regardless?

crookedneighbor commented:
Happens regardless. I should have just opened a pr into your branch, rather than creating a new PR. wasn't aware that I could do that until today.

crookedneighbor commented:
Following questions are from @hairlessbear:

> Regarding making the party page fetch on load, would #4816 be a problem? I'm worried that those in large parties would never see the invitation statuses of some members.

There should be a difference between fetching the party and fetching the recent messages. Might be a good time to revisit the fetch recent messages code and see if it needs to be adjusted.

> Regarding putting the accept/reject buttons on the party page, if #5157 is going to be merged, we should probably add that functionality there instead of here, right?

I don't think it necessarily needs to go there, but adding it here would mitigate the need for a migration script, since the few users that get a quest request in between deploys would still be able to join the quest from the party page. 

> Also, since you've now written the tests that you asked for in #5239, should that behavior be lumped into this PR?

I'm not quite happy with where that feature is at, so let's leave it separate for now.

crookedneighbor commented:
Actually, I've changed my mind. While we're cleaning up all these things, let's add #5239 (make quest completion modal appear before quest invite modal) to the list of nice to haves (see original post at top of PR)

hairlessbear commented:
> There should be a difference between fetching the party and fetching the recent messages. Might be a good time to revisit the fetch recent messages code and see if it needs to be adjusted.

I don't think there is a difference currently (I may be misunderstanding the code). `Fetch recent messages` calls [`sync(group)`](https://github.com/HabitRPG/habitrpg/blob/develop/website/views/options/social/chat-box.jade#L22), which I'm assuming means the whole party is fetched.

> I don't think it necessarily needs to go there, but adding it here would mitigate the need for a migration script, since the few users that get a quest request in between deploys would still be able to join the quest from the party page.

Avoiding the migration script is something I hadn't thought of; that's a good point. I suppose we could add accept/reject buttons to this PR, then once it's merged, I could rebase the other PR and integrate the new buttons. On the other hand, if #5157 gets merged first, we could rebase here and add the buttons. Both options require approximately the same amount of effort.

hairlessbear commented:
Regarding the `fetch recent messages` issue, [it looks like someone started working on it a while ago](https://github.com/HabitRPG/habitrpg/blob/develop/website/public/js/services/groupServices.js#L75-90), but left it as a "to-do".

crookedneighbor commented:
@hairlessbear Good work on making the quest completion modal appear before the next invite modal. :+1: Works really well now.

hairlessbear commented:
Thanks!

Would you prefer that a migration script be written for this, or the accept/reject buttons added to the party page? Only one is necessary, but both are currently in the checklists at the top.

crookedneighbor commented:
Do the accept/reject buttons. I'll work with Alys to do the migration script.

hairlessbear commented:
@crookedneighbor Will do.

crookedneighbor commented:
@hairlessbear Where are you at on this? Should I take it over?

crookedneighbor commented:
(Just itchy to close it, because I want it to be live :smile: )

hairlessbear commented:
Sorry, I won't be able to do it until tomorrow. I won't be offended if you
finish it :)

On Sat, Jun 13, 2015, 7:21 PM Blade Barringer notifications@github.com
wrote:

> (Just itchy to close it, because I want it to be live [image: :smile:] )
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/pull/5298#issuecomment-111759876.

crookedneighbor commented:
Just opened a PR to bring this branch up to date with develop: https://github.com/hairlessbear/habitrpg/pull/6

hairlessbear commented:
Thanks for doing that. I'll get to work on the accept/reject buttons now.

hairlessbear commented:
I'm getting an annoying synchronization issue. Each button has its action (`party.$questAccept`/`party.$questReject`) and `User.sync()` attached to it. The problem is that `User.sync()` finishes before `party.$questAccept`/`party.$questReject`, causing the accept/reject buttons to not hide until something else happens that causes `User.sync()` to run again. 

I haven't yet found a method to run commands attached to a single `ng-click` asynchronously; I'm pretty sure one doesn't exist. I'm still working on how exactly to make `User.sync()` not run until `party.$questAccept`/`party.$questReject` finish. Or possibly some other solution to fix the issue.

crookedneighbor commented:
You could just have a temporary variable on scope that hides the button when one of them is clicked. Or manually set the status of RSVPNeeded on the user.

hairlessbear commented:
I thought of those, but they feel so messy :confused:. But they would work, and the second one in particular is simple and intuitive. I'll probably just do that, assuming I don't think of something brilliant in the next minute or two.

hairlessbear commented:
Didn't come up with anything brilliant, and buttons now show/hide as expected. I'm not quite satisfied with how they look though, and I can't quite put my finger on why.
![selection_027](https://cloud.githubusercontent.com/assets/7059890/8151001/c09b672a-12cd-11e5-9d50-9ea07c8d925f.png)

crookedneighbor commented:
I expected them to be at the bottom, in the place where the "begin" and "cancel" buttons would be if you're the quest leader.

crookedneighbor commented on website/public/js/controllers/groupsCtrl.js:
```diff
@@ -579,17 +578,24 @@ habitrpg.controller("GroupsCtrl", ['$scope', '$rootScope', 'Shared', 'Groups', '
       }
 
+      $scope.questAccept = function(party){
+        Groups.questAccept(party);
```
Why not add `User.set({'party.quest.RSVPNeeded': false});` here and at questReject?

crookedneighbor replied:
_doesn't know if that will work_

hairlessbear replied:
I'll try it out and see what happens!

crookedneighbor replied:
Just tried it, and it seemed to work for me. 

Note, you'll have to change the buttons to call `ng-click='questAccept(party)'` and `ng-click='questReject(party)'` for it to work. 

We'll also probably want to add a test that the value gets updated.

hairlessbear replied:
This is interesting: just replacing the button actions with `questAccept(party)` and `questReject(party)` made it work properly, _without_ adding the `User.set` lines.

crookedneighbor replied:
¯_(ツ)_/¯

hairlessbear replied:
The code is now functional _and_ clean; it makes me so happy :smile:

hairlessbear commented:
Huh, I didn't think of that. I put them at the top to maximize visibility; if they're at the bottom, the user will have to scroll to see them. But they'd certainly look more in place there. I'll move them.

hairlessbear commented:
Oh man, you were so right about the button placement, this looks way better (and makes more sense too, since there's now explanatory text next to the buttons).

![selection_028](https://cloud.githubusercontent.com/assets/7059890/8151099/c8e292a6-12d1-11e5-9aab-1386f39c0515.png)

Alys commented:
Will these buttons now be the only way that people accept/reject quests? If so, I'd like them to be near the top of the screen so that they can be seen more easily (even if we have to move the instructions up too). Scrolling down within a box on the page will be too obscure for some users. It's will also be irritating for users in very large parties.

hairlessbear commented:
@Alys no no no, this is in addition to the quest invite modal. We wouldn't pull a fast one like that :smile:

Alys commented:
Fair enough, sorry. :)

I still preferred your buttons at the top though. :) But maybe we can look into that in a later change.

hairlessbear commented:
@crookedneighbor What do you think? Although I still think it's more visually appealing at the bottom (the buttons at the top are sized differently than the tabs, which bothers me way more than it should), I think it might be functionally better at the top, even if it's a bit uglier.

crookedneighbor commented:
Maybe the whole Ian block should be moved up and put before the quest participants list? (on the non-active quest only)

The important thing, at that point in the quest flow, are the instructions that Ian gives. 

Not sure though. For now, let's leave them where they are and revisit it later.

crookedneighbor commented:
Not going to worry about changing the "Fetch Recent Messages" button to only fetch messages, that's something that can be changed in a future refactoring. 

Just need to add a test for the watch and a migration script (with a test?). I'll handle both of those things, since they can be tricky, unless you'd like to take a stab at it.

hairlessbear commented:
I really want to do the migration script, but last time I tried, I got stuck. I figured out how to retrieve the users that were in parties with pending quests without issue (using `party.quest.key` and `party.quest.active`), but I couldn't figure out how to do the next step (getting users with status `undefined` and updating their `RSVPNeeded` to `true`). I could retrieve `party.quest.members`, but I couldn't figure out how to get the user IDs out of it to use in the necessary `User` updates.

It would probably be faster if you wrote the script, and I'll learn from reading it.

EDIT: And you should definitely do the testing. I still haven't taken the time to learn how the testing environment works, so it would take me a really long time to write the tests.

crookedneighbor commented:
I've never written one, so I'll probably need Alys to help me out. :smile:

hairlessbear commented:
I've noticed an interesting side-effect of these changes, and I'm not quite sure what to do about it.

Now, if an invite is sent out, and an invited user hits "Fetch Recent Messages", the quest information will display on the party page, but the accept/reject buttons won't show, and the invite modal won't pop-up. This of course makes sense, since these are controlled by `RSVPNeeded`, which isn't updated on group sync. 

This behavior differs than the current behavior (i.e., "Fetch Recent Messages" causes the modal to pop-up), and I think it may lead to some confusion and/or frustration. It seems strange to load some of the quest information on group sync, but not all of it. 

I haven't come up with a great solution to this yet :confused: Is there a way to update `RSVPNeeded` when "Fetch Recent Messages" is pushed, without running a full `User.sync()`?

Alys commented on website/public/js/controllers/notificationCtrl.js:
```diff
@@ -154,9 +154,9 @@ habitrpg.controller('NotificationCtrl',
-    $rootScope.$watch('party.quest.key && !party.quest.active && party.quest.members[user._id] == undefined', function(after, before){
-      if (after == before || after != true) return;
-      $rootScope.openModal('questInvitation');
+    $rootScope.$watch('user.party.quest.RSVPNeeded && (user.party.quest.completed == "" || user.party.quest.completed == undefined)', function(after, before){
```
For `(user.party.quest.completed == "" || user.party.quest.completed == undefined)`, is it better to put the undefined case first? Will `== ""` throw an error or warning if `user.party.quest.completed` does not exist?

hairlessbear replied:
I don't think the order matters. When I tested this, I tested both cases of `user.party.quest.completed` (when it equals `""` and when it's null), and they both worked as expected.

Alys commented:
For the migration script, if I understand correctly, you want to find all users who exist in a party's `quest.members` object with a value of `null` (they have not accepted nor rejected the quest), and you want to give those players the `party.quest.RSVPNeeded` attribute with a value of true.

If that's not correct, ignore everything below. :)

For this kind of migration, by far the easiest way to do it is to capture all quest.members for all parties with inactive quests into an `output` file (script takes about 10 to 20 seconds to run on the production database):

```
db.groups.find( {
    'type': 'party',
    'quest.active': false,
    'quest.key': {
        $exists:true,
        $ne:null,
        $ne:"",
    },
}, {
    'quest.members':1
}).pretty();
```

`output` then contains data like this:

```
{
    "_id" : "12340a62-4a34-4cd5-affe-3e11099a6082",
    "quest" : {
        "members" : {
            "1234fd38-9470-4893-84e0-0c62cd9b4ef7" : true,
            "1234a067-32d8-4be5-8bb8-eb4e526ad8ed" : null
        },
    },
}
{
  ... next party ....
}
```

Then use grep to get the members with a value of null, use sed to strip everything from those lines that isn't a UUID, and save the UUIDs to a temp file:
`cat output | grep null | sed -e 's/" : .*/",/' > /tmp/t`

then read the temp file into a script that sets the required attribute (script takes 10 to 20 seconds on a local copy of the production database, and probably slightly longer on the real one):

```
var uuidArray=[
  ... UUIDS go here ...
];
db.users.update({ '_id':{$in: uuidArray} }, {$set: {
    'party.quest.RSVPNeeded': true
}},{ multi: true });
```

The whole process is done in under a minute. 

You risk incorrectly setting party.quest.RSVPNeeded to true for any players who accept a quest within that minute, but it's a low risk, and we can check for that having happened afterwards with another couple of database queries. Even with that final check, this is much faster than writing, testing, and debugging an all-in-one script.

hairlessbear commented:
@Alys That is extremely useful; thank you! The trouble I had was definitely caused by trying to do it all in one script. Splitting it up makes a lot of sense, and using `grep` + `sed` to extract the UIDs is something I hadn't even considered. Very cool.

I'm still trying to figure out how to do the specifics of what you've described (e.g., pipe the output of the first query to a file), but logically what you've written makes complete sense. Thanks again.

Alys commented:
This is how we execute migration scripts, with the output being sent to a file (that part is optional of course):
`mongo ds031379-a0.mongolab.com:31379/habitrpg -u mongousername -p nopenopenope < /path/to/script > output`

hairlessbear commented:
Thanks! In retrospect, that makes perfect sense, but I suppose most things do once they're explained :smile:

hairlessbear committed:
Merge branch 'crookedneighbor-quest-invite-modal-on-sync' into quest_invite_modal_on_user_sync

hairlessbear committed:
Changed completion modal watch to be on  instead of

hairlessbear commented:
@crookedneighbor Since switching the invite watch to `$scope` worked, and it relies on `user.party.quest.completed`, I figured that the completion modal watch would also be safe to move to `$scope`. I tested it and it works as expected, but of course let me know if I missed something that makes this a bad idea.

hairlessbear commented:
Whoops, I suppose tests failing are an example of a problem. I'm looking into it now

hairlessbear commented:
I see the problem; you used [`expect(rootScope.openModal).to.not.be.called;`](https://github.com/hairlessbear/habitrpg/blob/4fc04fea24889d3550cfb01f508466ff56c5ad40/test/spec/controllers/notificationCtrlSpec.js#L45) to test whether the quest invite modal is opened when it shouldn't be in multiple situations. It's still not being opened, but the check is catching the completion modal, making the test fail.

I think I know how to fix this. Gimme a minute.

hairlessbear committed:
Changed 'invite modal does not open' tests to focus on completion modal opening, but being the only modal to open. Also modified the named of two tests to be more intuitive and accurate

hairlessbear commented:
~~@crookedneighbor What do you think? The various "quest invitation modal doesn't open" tests now check that `rootScope.openModal` is called once, and that it's called with `questCompleted`. This implies that the invite modal hasn't been opened, so I think it's effective, even if it's a bit indirect.~~

Edit: See below for updated message

hairlessbear committed:
Added quest completion modal watch tests

hairlessbear committed:
Modified 'RSVPNeeded is not true' test to not set user.party.quest.completed because it is not needed for the test

hairlessbear committed:
Removed duplicated test

hairlessbear commented:
@crookedneighbor Summary of the most recent changes:
1. Modified the `does not open quest invitation modal if RSVPNeeded is not true` test to not set `user.party.quest.completed` because it's not needed.
2. Removed one of the `does not open quest invitation modal when quest.completed is X` tests. There were originally two, one for when `quest.completed` is true and one for when `quest.completed` is false. After I told you about the values of `quest.completed`, you modified the tests to use quest keys, making them duplicates. I also modified the remaining test's name to be accurate (`does not open quest invitation modal if quest.completed contains a quest key`)
3. Modified the `does not open quest invitation modal if quest.completed contains a quest key` test to test for the quest completion modal to be opened, and to be the only modal that's opened, implying that the invitation modal hasn't been opened.
4. Added tests for the quest completion watch, because why not?

crookedneighbor commented on test/spec/controllers/notificationCtrlSpec.js:
```diff
@@ -0,0 +1,78 @@
+      expect(rootScope.openModal).to.not.be.called;
+    });
+
+    it('does not open quest invitation modal if quest.completed contains a quest key', function() {
```
Good catch.

crookedneighbor commented on test/spec/controllers/notificationCtrlSpec.js:
```diff
@@ -0,0 +1,78 @@
+      expect(rootScope.openModal).to.be.calledWith('questCompleted', {controller:'InventoryCtrl'});
+    });
+  });
+  describe('Quest Completion modal watch', function() {
```
Wonderful! Add a line break just above this describe block.

crookedneighbor commented on test/spec/controllers/notificationCtrlSpec.js:
```diff
@@ -0,0 +1,78 @@
+
+      // Completion modal should open, but should be the only modal that opens
+      expect(rootScope.openModal).to.be.calledOnce;
+      expect(rootScope.openModal).to.be.calledWith('questCompleted', {controller:'InventoryCtrl'});
```
This should probably be:

```
expect(rootScope.openModal).to.not.be.calledWith('questInvitation', {controller:'InventoryCtrl'});
```

hairlessbear replied:
Ah, that's much more intuitive. I'll go ahead and change it, but I just realized that we never explicitly check for the completion modal opening before the invite modal, and this check is testing that behavior. I'll make the suggested change and then move this test to the challenge watch block with a different title.

hairlessbear committed:
Added a line break

crookedneighbor commented on test/spec/controllers/notificationCtrlSpec.js:
```diff
@@ -0,0 +1,78 @@
+      scope.$digest();
+
+      // Completion modal should open, but should be the only modal that opens
+      expect(rootScope.openModal).to.be.calledOnce;
```
If you make the change I suggested below, just remove this line and the comment.

hairlessbear committed:
Added quest completion modal test for when RSVPNeeded is true; modified quest invite test to use 'to.not.be.calledWith'

hairlessbear committed:
Added explanatory comment

hairlessbear commented:
@crookedneighbor @Alys What are your thoughts on the [party sync issue](https://github.com/HabitRPG/habitrpg/pull/5298#issuecomment-112294341) I mentioned earlier? Since that's a relatively specific, possibly fringe case, are we just going to not worry about it?

crookedneighbor commented:
I'd say let's not worry about it.

crookedneighbor commented:
I've merged this into a branch (I've also brought it up to date with develop): https://github.com/HabitRPG/habitrpg/tree/feature_quest_invite_modal_on_user_sync

@Alys When we're ready to merge in and deploy it, would you mind taking care of the migration step?

Alys commented:
Yes, happy to do the migration. I assume it's still the same as described in my comment https://github.com/HabitRPG/habitrpg/pull/5298#issuecomment-112359954

crookedneighbor commented:
Yep, you got it. I'm ready to merge and deploy at any time today, so just let me know when you're ready.

crookedneighbor commented:
Notated toward your next tier. About to deploy to prod and run the migration

hairlessbear commented:
Yay! I'm glad this is finally getting added.

Alys commented:
migration done
